File: packages/swift-agent-protocol/Sources/OpenPawProtocol/Event.swift
import Foundation

// MARK: - Body

/// The `type`/`payload` pair of a normalized event, decoded into a typed case.
///
/// A `type` this build does not know decodes into `.unsupported`, which re-encodes
/// byte-for-byte. Forward compatibility is a protocol requirement: a phone running an
/// older build must not drop or corrupt events emitted by a newer host.
public enum Body: Sendable, Hashable {
    case agentStarted(AgentLifecycle)
    case agentWorking(AgentLifecycle)
    case agentCompleted(AgentLifecycle)
    case agentFailed(AgentLifecycle)
    case turnStarted(TurnStarted)
    case turnDelta(TurnDelta)
    case turnCompleted(TurnCompleted)
    case toolStarted(ToolStarted)
    case toolOutput(ToolOutput)
    case toolCompleted(ToolCompleted)
    case toolFailed(ToolFailed)
    case permissionRequested(PermissionRequested)
    case permissionResolved(PermissionResolved)
    case questionRequested(QuestionRequested)
    case questionAnswered(QuestionAnswered)
    case planCreated(Plan)
    case planUpdated(Plan)
    case fileRead(FileChange)
    case fileCreated(FileChange)
    case fileModified(FileChange)
    case fileDeleted(FileChange)
    case usageUpdated(UsageUpdated)
    case contextUpdated(ContextUpdated)
    case unsupported(type: String, payload: JSONValue)

    /// The known event type, or `nil` for `.unsupported`.
    public var kind: EventType? {
        switch self {
        case .agentStarted: .agentStarted
        case .agentWorking: .agentWorking
        case .agentCompleted: .agentCompleted
        case .agentFailed: .agentFailed
        case .turnStarted: .turnStarted
        case .turnDelta: .turnDelta
        case .turnCompleted: .turnCompleted
        case .toolStarted: .toolStarted
        case .toolOutput: .toolOutput
        case .toolCompleted: .toolCompleted
        case .toolFailed: .toolFailed
        case .permissionRequested: .permissionRequested
        case .permissionResolved: .permissionResolved
        case .questionRequested: .questionRequested
        case .questionAnswered: .questionAnswered
        case .planCreated: .planCreated
        case .planUpdated: .planUpdated
        case .fileRead: .fileRead
        case .fileCreated: .fileCreated
        case .fileModified: .fileModified
        case .fileDeleted: .fileDeleted
        case .usageUpdated: .usageUpdated
        case .contextUpdated: .contextUpdated
        case .unsupported: nil
        }
    }

    /// The wire value of `type`, defined for every case including `.unsupported`.
    public var typeName: String {
        if case .unsupported(let type, _) = self { return type }
        // `kind` is non-nil for every case except `.unsupported`.
        return kind!.rawValue
    }
}

// MARK: - Event

public struct Event: Codable, Sendable, Hashable, Identifiable {
    public var version: String
    public var eventID: EventID
    public var sessionID: SessionID
    public var agent: AgentKind
    public var seq: UInt64
    public var timestamp: Date
    public var cwd: String?
    public var gitBranch: String?
    public var multiplexerTarget: String?
    public var body: Body

    public var id: EventID { eventID }

    /// The known event type, or `nil` when the host emitted a type this build predates.
    public var kind: EventType? { body.kind }

    public init(
        version: String = OpenPawCoding.version,
        eventID: EventID,
        sessionID: SessionID,
        agent: AgentKind,
        seq: UInt64 = 0,
        timestamp: Date,
        cwd: String? = nil,
        gitBranch: String? = nil,
        multiplexerTarget: String? = nil,
        body: Body
    ) {
        self.version = version
        self.eventID = eventID
        self.sessionID = sessionID
        self.agent = agent
        self.seq = seq
        self.timestamp = timestamp
        self.cwd = cwd
        self.gitBranch = gitBranch
        self.multiplexerTarget = multiplexerTarget
        self.body = body
    }

    /// Builds an event with the same content addressed identifier the host derives.
    public init(
        session: SessionID,
        agent: AgentKind,
        sourceKey: String,
        timestamp: Date,
        body: Body
    ) {
        self.init(
            eventID: EventID(session: session, sourceKey: sourceKey),
            sessionID: session,
            agent: agent,
            seq: 0,
            timestamp: timestamp,
            body: body
        )
    }

    public func withSeq(_ seq: UInt64) -> Event {
        var copy = self
        copy.seq = seq
        return copy
    }

    public func withContext(cwd: String?, gitBranch: String?) -> Event {
        var copy = self
        copy.cwd = cwd
        copy.gitBranch = gitBranch
        return copy
    }

    // MARK: Coding

    enum CodingKeys: String, CodingKey {
        case version
        case eventID = "event_id"
        case sessionID = "session_id"
        case agent
        case seq
        case timestamp
        case cwd
        case gitBranch = "git_branch"
        case multiplexerTarget = "multiplexer_target"
        case type
        case payload
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        version = try container.decode(String.self, forKey: .version)
        eventID = try container.decode(EventID.self, forKey: .eventID)
        sessionID = try container.decode(SessionID.self, forKey: .sessionID)
        agent = try container.decode(AgentKind.self, forKey: .agent)
        seq = try container.decode(UInt64.self, forKey: .seq)
        timestamp = try container.decode(Date.self, forKey: .timestamp)
        cwd = try container.decodeIfPresent(String.self, forKey: .cwd)
        gitBranch = try container.decodeIfPresent(String.self, forKey: .gitBranch)
        multiplexerTarget = try container.decodeIfPresent(String.self, forKey: .multiplexerTarget)

        let type = try container.decode(String.self, forKey: .type)
        func payload<T: Decodable>(_ type: T.Type) throws -> T {
            try container.decode(T.self, forKey: .payload)
        }
        switch EventType(rawValue: type) {
        case .agentStarted: body = .agentStarted(try payload(AgentLifecycle.self))
        case .agentWorking: body = .agentWorking(try payload(AgentLifecycle.self))
        case .agentCompleted: body = .agentCompleted(try payload(AgentLifecycle.self))
        case .agentFailed: body = .agentFailed(try payload(AgentLifecycle.self))
        case .turnStarted: body = .turnStarted(try payload(TurnStarted.self))
        case .turnDelta: body = .turnDelta(try payload(TurnDelta.self))
        case .turnCompleted: body = .turnCompleted(try payload(TurnCompleted.self))
        case .toolStarted: body = .toolStarted(try payload(ToolStarted.self))
        case .toolOutput: body = .toolOutput(try payload(ToolOutput.self))
        case .toolCompleted: body = .toolCompleted(try payload(ToolCompleted.self))
        case .toolFailed: body = .toolFailed(try payload(ToolFailed.self))
        case .permissionRequested: body = .permissionRequested(try payload(PermissionRequested.self))
        case .permissionResolved: body = .permissionResolved(try payload(PermissionResolved.self))
        case .questionRequested: body = .questionRequested(try payload(QuestionRequested.self))
        case .questionAnswered: body = .questionAnswered(try payload(QuestionAnswered.self))
        case .planCreated: body = .planCreated(try payload(Plan.self))
        case .planUpdated: body = .planUpdated(try payload(Plan.self))
        case .fileRead: body = .fileRead(try payload(FileChange.self))
        case .fileCreated: body = .fileCreated(try payload(FileChange.self))
        case .fileModified: body = .fileModified(try payload(FileChange.self))
        case .fileDeleted: body = .fileDeleted(try payload(FileChange.self))
        case .usageUpdated: body = .usageUpdated(try payload(UsageUpdated.self))
        case .contextUpdated: body = .contextUpdated(try payload(ContextUpdated.self))
        case nil:
            let raw = try container.decodeIfPresent(JSONValue.self, forKey: .payload) ?? .object([:])
            body = .unsupported(type: type, payload: raw)
        }
    }

    public func encode(to encoder: any Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        // Key order matches the host: envelope, then type, then payload. The three
        // context keys are always present, encoded as null when absent.
        try container.encode(version, forKey: .version)
        try container.encode(eventID, forKey: .eventID)
        try container.encode(sessionID, forKey: .sessionID)
        try container.encode(agent, forKey: .agent)
        try container.encode(seq, forKey: .seq)
        try container.encode(timestamp, forKey: .timestamp)
        try encodeNullable(cwd, forKey: .cwd, into: &container)
        try encodeNullable(gitBranch, forKey: .gitBranch, into: &container)
        try encodeNullable(multiplexerTarget, forKey: .multiplexerTarget, into: &container)
        try container.encode(body.typeName, forKey: .type)

        switch body {
        case .agentStarted(let value), .agentWorking(let value), .agentCompleted(let value),
            .agentFailed(let value):
            try container.encode(value, forKey: .payload)
        case .turnStarted(let value): try container.encode(value, forKey: .payload)
        case .turnDelta(let value): try container.encode(value, forKey: .payload)
        case .turnCompleted(let value): try container.encode(value, forKey: .payload)
        case .toolStarted(let value): try container.encode(value, forKey: .payload)
        case .toolOutput(let value): try container.encode(value, forKey: .payload)
        case .toolCompleted(let value): try container.encode(value, forKey: .payload)
        case .toolFailed(let value): try container.encode(value, forKey: .payload)
        case .permissionRequested(let value): try container.encode(value, forKey: .payload)
        case .permissionResolved(let value): try container.encode(value, forKey: .payload)
        case .questionRequested(let value): try container.encode(value, forKey: .payload)
        case .questionAnswered(let value): try container.encode(value, forKey: .payload)
        case .planCreated(let value), .planUpdated(let value):
            try container.encode(value, forKey: .payload)
        case .fileRead(let value), .fileCreated(let value), .fileModified(let value),
            .fileDeleted(let value):
            try container.encode(value, forKey: .payload)
        case .usageUpdated(let value): try container.encode(value, forKey: .payload)
        case .contextUpdated(let value): try container.encode(value, forKey: .payload)
        case .unsupported(_, let payload): try container.encode(payload, forKey: .payload)
        }
    }

    private func encodeNullable(
        _ value: String?,
        forKey key: CodingKeys,
        into container: inout KeyedEncodingContainer<CodingKeys>
    ) throws {
        if let value {
            try container.encode(value, forKey: key)
        } else {
            try container.encodeNil(forKey: key)
        }
    }
}

File: host/crates/openpaw-protocol/src/event.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::PROTOCOL_VERSION;
use crate::agent::AgentKind;
use crate::ids::{EventId, SessionId};
use crate::payload::{
    AgentLifecycle, ContextUpdated, FileChange, PermissionRequested, PermissionResolved, Plan,
    QuestionAnswered, QuestionRequested, ToolCompleted, ToolFailed, ToolOutput, ToolStarted,
    TurnCompleted, TurnDelta, TurnStarted, UsageUpdated,
};
use crate::wire_enum::wire_enum;

wire_enum! {
    /// Discriminant of [`Body`], mirroring `event.schema.json#/$defs/eventType`.
    pub enum EventType {
        /// A session was discovered or launched.
        AgentStarted = "agent.started",
        /// The agent began or resumed producing output.
        AgentWorking = "agent.working",
        /// The agent finished its turn and is idle.
        AgentCompleted = "agent.completed",
        /// The agent stopped because of an error.
        AgentFailed = "agent.failed",
        /// A conversation turn started.
        TurnStarted = "turn.started",
        /// A fragment of a streaming turn.
        TurnDelta = "turn.delta",
        /// A conversation turn finished.
        TurnCompleted = "turn.completed",
        /// A tool call started.
        ToolStarted = "tool.started",
        /// A chunk of tool output.
        ToolOutput = "tool.output",
        /// A tool call succeeded.
        ToolCompleted = "tool.completed",
        /// A tool call failed.
        ToolFailed = "tool.failed",
        /// The agent is blocked on an approval decision.
        PermissionRequested = "permission.requested",
        /// An approval decision was recorded.
        PermissionResolved = "permission.resolved",
        /// The agent is blocked on a question.
        QuestionRequested = "question.requested",
        /// A question was answered.
        QuestionAnswered = "question.answered",
        /// A plan was published.
        PlanCreated = "plan.created",
        /// A published plan changed.
        PlanUpdated = "plan.updated",
        /// A file was read.
        FileRead = "file.read",
        /// A file was created.
        FileCreated = "file.created",
        /// A file was modified.
        FileModified = "file.modified",
        /// A file was deleted.
        FileDeleted = "file.deleted",
        /// Token and cost counters changed.
        UsageUpdated = "usage.updated",
        /// Context window occupancy changed.
        ContextUpdated = "context.updated",
    }
}

/// Discriminated payload of an [`Event`].
///
/// Serialized adjacently as `{"type": ..., "payload": {...}}` and flattened
/// into the envelope, which is exactly the shape the JSON schema requires.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum Body {
    /// See [`EventType::AgentStarted`].
    #[serde(rename = "agent.started")]
    AgentStarted(AgentLifecycle),
    /// See [`EventType::AgentWorking`].
    #[serde(rename = "agent.working")]
    AgentWorking(AgentLifecycle),
    /// See [`EventType::AgentCompleted`].
    #[serde(rename = "agent.completed")]
    AgentCompleted(AgentLifecycle),
    /// See [`EventType::AgentFailed`].
    #[serde(rename = "agent.failed")]
    AgentFailed(AgentLifecycle),
    /// See [`EventType::TurnStarted`].
    #[serde(rename = "turn.started")]
    TurnStarted(TurnStarted),
    /// See [`EventType::TurnDelta`].
    #[serde(rename = "turn.delta")]
    TurnDelta(TurnDelta),
    /// See [`EventType::TurnCompleted`].
    #[serde(rename = "turn.completed")]
    TurnCompleted(TurnCompleted),
    /// See [`EventType::ToolStarted`].
    #[serde(rename = "tool.started")]
    ToolStarted(ToolStarted),
    /// See [`EventType::ToolOutput`].
    #[serde(rename = "tool.output")]
    ToolOutput(ToolOutput),
    /// See [`EventType::ToolCompleted`].
    #[serde(rename = "tool.completed")]
    ToolCompleted(ToolCompleted),
    /// See [`EventType::ToolFailed`].
    #[serde(rename = "tool.failed")]
    ToolFailed(ToolFailed),
    /// See [`EventType::PermissionRequested`].
    #[serde(rename = "permission.requested")]
    PermissionRequested(PermissionRequested),
    /// See [`EventType::PermissionResolved`].
    #[serde(rename = "permission.resolved")]
    PermissionResolved(PermissionResolved),
    /// See [`EventType::QuestionRequested`].
    #[serde(rename = "question.requested")]
    QuestionRequested(QuestionRequested),
    /// See [`EventType::QuestionAnswered`].
    #[serde(rename = "question.answered")]
    QuestionAnswered(QuestionAnswered),
    /// See [`EventType::PlanCreated`].
    #[serde(rename = "plan.created")]
    PlanCreated(Plan),
    /// See [`EventType::PlanUpdated`].
    #[serde(rename = "plan.updated")]
    PlanUpdated(Plan),
    /// See [`EventType::FileRead`].
    #[serde(rename = "file.read")]
    FileRead(FileChange),
    /// See [`EventType::FileCreated`].
    #[serde(rename = "file.created")]
    FileCreated(FileChange),
    /// See [`EventType::FileModified`].
    #[serde(rename = "file.modified")]
    FileModified(FileChange),
    /// See [`EventType::FileDeleted`].
    #[serde(rename = "file.deleted")]
    FileDeleted(FileChange),
    /// See [`EventType::UsageUpdated`].
    #[serde(rename = "usage.updated")]
    UsageUpdated(UsageUpdated),
    /// See [`EventType::ContextUpdated`].
    #[serde(rename = "context.updated")]
    ContextUpdated(ContextUpdated),
}

impl Body {
    /// The discriminant of this payload.
    pub const fn kind(&self) -> EventType {
        match self {
            Body::AgentStarted(_) => EventType::AgentStarted,
            Body::AgentWorking(_) => EventType::AgentWorking,
            Body::AgentCompleted(_) => EventType::AgentCompleted,
            Body::AgentFailed(_) => EventType::AgentFailed,
            Body::TurnStarted(_) => EventType::TurnStarted,
            Body::TurnDelta(_) => EventType::TurnDelta,
            Body::TurnCompleted(_) => EventType::TurnCompleted,
            Body::ToolStarted(_) => EventType::ToolStarted,
            Body::ToolOutput(_) => EventType::ToolOutput,
            Body::ToolCompleted(_) => EventType::ToolCompleted,
            Body::ToolFailed(_) => EventType::ToolFailed,
            Body::PermissionRequested(_) => EventType::PermissionRequested,
            Body::PermissionResolved(_) => EventType::PermissionResolved,
            Body::QuestionRequested(_) => EventType::QuestionRequested,
            Body::QuestionAnswered(_) => EventType::QuestionAnswered,
            Body::PlanCreated(_) => EventType::PlanCreated,
            Body::PlanUpdated(_) => EventType::PlanUpdated,
            Body::FileRead(_) => EventType::FileRead,
            Body::FileCreated(_) => EventType::FileCreated,
            Body::FileModified(_) => EventType::FileModified,
            Body::FileDeleted(_) => EventType::FileDeleted,
            Body::UsageUpdated(_) => EventType::UsageUpdated,
            Body::ContextUpdated(_) => EventType::ContextUpdated,
        }
    }
}

/// One normalized observation about an agent session.
///
/// Field order matters: it is the on-the-wire key order both protocol
/// implementations emit.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Event {
    /// Always [`PROTOCOL_VERSION`].
    pub version: String,
    /// Content addressed identifier, stable across re-reads of the source.
    pub event_id: EventId,
    /// Session this event belongs to.
    pub session_id: SessionId,
    /// Agent that produced the source data.
    pub agent: AgentKind,
    /// Monotonic per session, assigned in source order.
    pub seq: u64,
    /// When the source recorded the observation.
    #[serde(with = "time::serde::rfc3339")]
    pub timestamp: OffsetDateTime,
    /// Working directory of the session, when known.
    #[serde(default)]
    pub cwd: Option<String>,
    /// Git branch checked out in `cwd`, when known.
    #[serde(default)]
    pub git_branch: Option<String>,
    /// Multiplexer target the session lives in, e.g. a tmux `work:2.0`.
    #[serde(default)]
    pub multiplexer_target: Option<String>,
    /// Type discriminant and payload.
    #[serde(flatten)]
    pub body: Body,
}

impl Event {
    /// Builds an event with `seq` 0 and no context.
    ///
    /// `source_key` must uniquely name the origin of this event inside the
    /// session so that [`EventId::derive`] makes re-ingestion idempotent.
    pub fn new(
        session: &SessionId,
        agent: AgentKind,
        source_key: &str,
        timestamp: OffsetDateTime,
        body: Body,
    ) -> Event {
        Event {
            version: PROTOCOL_VERSION.to_owned(),
            event_id: EventId::derive(session, source_key),
            session_id: session.clone(),
            agent,
            seq: 0,
            timestamp,
            cwd: None,
            git_branch: None,
            multiplexer_target: None,
            body,
        }
    }

    /// The type discriminant of this event.
    pub const fn kind(&self) -> EventType {
        self.body.kind()
    }

    /// Assigns the per-session sequence number.
    pub fn with_seq(mut self, seq: u64) -> Event {
        self.seq = seq;
        self
    }

    /// Attaches the working directory and git branch.
    pub fn with_context(mut self, cwd: Option<String>, git_branch: Option<String>) -> Event {
        self.cwd = cwd;
        self.git_branch = git_branch;
        self
    }

    /// Attaches the multiplexer target the session was discovered in.
    pub fn with_multiplexer_target(mut self, target: Option<String>) -> Event {
        self.multiplexer_target = target;
        self
    }
}
